encoding/ssz/query/buildTree.go

```diff
@@ -0,0 +1,284 @@
+				return fmt.Errorf("vector index %d: %w", i, err)
+			}
+		}
+	}
+
+	addPadding(w, length)
+	w.Commit(start)
+	return nil
+}
+
+// buildListSubtree merkleizes a variable-length sequence of elements.
+// Basic types are packed into 32-byte chunks; composite types are processed recursively.
+// The length is mixed into the root hash via CommitWithMixin.
+func buildListSubtree(info *SszInfo, v reflect.Value, w *ssz.Wrapper) error {
+	li, err := info.ListInfo()
+	if err != nil {
+		return err
+	}
+
+	length := v.Len()
+	limit := int(li.Limit())
+	start := w.Indx()
+
+	if li.element.sszType.isBasic() {
+		// Lists of basic types are packed into 32-byte chunks
+		packBasicElements(li.element, v, w, length)
+
+		limit = int(ssz.CalculateLimit(uint64(limit), uint64(length), uint64(itemLength(li.element))))
+	} else {
+		// General case: list of composite elements
+		for i := 0; i < length; i++ {
+			if err := buildTree(li.element, v.Index(i), w); err != nil {
+				return fmt.Errorf("list index %d: %w", i, err)
+			}
+		}
+	}
+
+	w.CommitWithMixin(start, length, int(nextPowerOfTwo(uint64(limit))))
+	return nil
+}
+
+// buildBitlistSubtree merkleizes a variable-length bit sequence.
+// The termination bit is cleared before chunking, and the bit length
+// is mixed into the root hash via CommitWithMixin.
+func buildBitlistSubtree(info *SszInfo, v reflect.Value, w *ssz.Wrapper) error {
+	bi, err := info.BitlistInfo()
+	if err != nil {
+		return err
+	}
+
+	bitlistBytes := v.Bytes()
+
+	// Handle zero-initialized bitlist: create a single byte with just the termination bit
+	if len(bitlistBytes) == 0 {
+		bitlistBytes = []byte{0x01}
+	}
+
+	// Use go-bitfield to get length and bytes with termination bit cleared
+	bl := bitfield.Bitlist(bitlistBytes)
+	data := bl.BytesNoTrim()
+
+	start := w.Indx()
+
+	// Add bytes in 32-byte chunks
+	for i := 0; i < len(data); i += 32 {
+		end := i + 32
+		if end > len(data) {
+			end = len(data)
+		}
+		w.AddBytes(data[i:end])
+	}
+
+	// Number of 32-byte chunks to cover the max bit capacity.
+	// In consensus specs these max bit sizes are chosen so this is a power of 2.
+	limitChunks := int((bi.limit + 255) / 256)
+
+	w.CommitWithMixin(start, int(bi.length), limitChunks)
+	return nil
+}
+
+// buildBitvectorSubtree merkleizes a fixed-length bit sequence.
+// Bytes are chunked into 32-byte leaves and padded to the next power of 2.
+// Unlike bitlists, no length mixin is applied since the size is fixed.
+func buildBitvectorSubtree(info *SszInfo, v reflect.Value, w *ssz.Wrapper) error {
+	bv, err := info.BitvectorInfo()
+	if err != nil {
+		return err
+	}
+
+	bitvectorBytes := v.Bytes()
+
+	// If bitvector is not initialized (zero-valued), create appropriately-sized zero array
+	if len(bitvectorBytes) == 0 {
+		expectedBytes := int((bv.length + 7) / 8) // ceil(bits/8)
+		bitvectorBytes = make([]byte, expectedBytes)
+	}
+
+	start := w.Indx()
+	// Add bytes in 32-byte chunks
+	numChunks := 0
```

Comment:
```suggestion
	length := bv.Length()
	numChunks := length / 32
```

I think we don't have to increment `numChunks` in the loop, just calculate with the length.

Reply:
Totally agree with this. I think I originally had something similar but ended up counting chunks in the loop because the bitvector was uninitialized (which is illegal by the SSZ spec https://github.com/ethereum/consensus-specs/blob/dev/ssz/simple-serialize.md#illegal-types), so I'm also making the following change:

```diff
	// If bitvector is not initialized (zero-valued), create appropriately-sized zero array
	if len(bitvectorBytes) == 0 {
-		expectedBytes := int((bv.length + 7) / 8) // ceil(bits/8)
-		bitvectorBytes = make([]byte, expectedBytes)
	}
	
	// bitvector must be initialized and non-empty
	if len(bitvectorBytes) == 0 {
+		return fmt.Errorf("bitvector field is uninitialized (nil or empty slice)")
	}
```